lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from opaque_keys.edx.locator import BlockUsageLocator
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from hashlib import sha256
+import json
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return self.hashed
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
```

Comment:
While unlikely, combining the key like this can lead to collisions and indeterminate behavior. You could just make the key be the tuple `(block.block_key, block.max_score, block.weight)` itself -- there's no need to format it into a string. I think the easiest thing might be to simply make `BlockRecord` inherit from [`namedtuple`](https://docs.python.org/2/library/collections.html#collections.namedtuple), in which case you'd get sorting and `_as_dict` behavior for free. You'll also want to be very specific about the formatting and sorting to make sure that you always get the same JSON representation for a given set of blocks.

If you did that `namedtuple` conversion, this code might look something like:

```python
blocks_json = json.dumps(
    [block._asdict() for block in sorted(blocks)],
    separators=(',', ':'),
    sort_keys=True
)
```